torch/_inductor/autotune_process.py

```diff
@@ -150,15 +148,130 @@ def close(self) -> None:
 
     def terminate(self) -> None:
         """
-        Signal the child process to terminate and wait for it to exit.
+        Signal the child process to terminate.
         """
         if self.process is not None:
             self.put(None)
+
+    def wait(self) -> None:
+        """
+        Wait for the child process to exit.
+        """
+        if self.process is not None:
             self.process.wait()
             self.close()
 
 
-tuning_process = TuningProcess()
+@dataclasses.dataclass
+class TuningProcessPool:
+    """
+    Maintains a pool of TuningProcesses to benchmark kernels in parallel
+    across devices. By default, we create one TuningProcess per device and
+    set the sub-process environment to make only that device visible.
+    """
+
+    processes: Optional[Queue[TuningProcess]] = None
+    executor: Optional[ThreadPoolExecutor] = None
+
+    def initialize(self, count: Optional[int] = None) -> None:
+        """
+        Start the child processes.
+        """
+        assert (self.processes is None) == (self.executor is None)
+        if self.processes is not None:
+            return
+
+        if config.autotune_multi_device:
+            count = count or torch.cuda.device_count()
+        else:
+            count = 1
+        assert count > 0 and count <= torch.cuda.device_count()
+
+        # Launch the child processes and push a msg to "warm up"
+        self.processes = Queue()
+        for device in range(count):
+            p = TuningProcess(device=device if config.autotune_multi_device else None)
```

Comment:
@masnesral @eellison I believe, there is a subtlety here worth revisiting.

Above we iterate over the `range(count)` and pass the `device` counter variable downstream, which is then set as the value of the `CUDA_VISIBLE_DEVICES` env variable when launching the subprocess. Now if the user has set `CUDA_VISIBLE_DEVICES=3,4,5` for the parent process, the `torch.cuda.device_count()` above will be 3, and we'll initialize the 3 subprocesses with `CUDA_VISIBLE_DEVICES=0`, `=1`, and `=2`. I guess, this is not what the user would expect?

In AITemplate, we actually parse the `CUDA_VISIBLE_DEVICES` env variable (fetched by `target.dev_select_flag()` [here](https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/compiler/compiler.py#L270-L275)) to detect the user-specified devices. Perhaps we should do something similar here, too? Or maybe there is a `torch.cuda.<something>` to detect the visible device IDs (didn't find from a quick check)?

Reply:
@aakhundov, ah sure. That sounds like a better behavior. Lemme work on a follow-up.